feat: LEAP-1454: Custom action buttons in LSF

hlomzik opened the pull request:
Introducing custom action buttons for LSF to allow additional actions, controlled from outside of editor. Provided buttons will completely replace the internal ones, but there is a way to reuse buttons you need, just add the button with internal name and no extra params.

TODO: Add example

Custom buttons need to behave similarly to internal ones, so we are using flags and additional checks. They will call the same event `handleCustomButton(button_name)` so external app will handle `customButton` event with `store, button_name, options` params.

Also new types for global MST stores are introduced in a draft state with intention to expand them with more props and methods so they can be used in other TypeScript code, which currently uses `any` for any MST stuff.

### PR fulfills these requirements
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [x] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

### Describe the reason for change
We need to introduce new buttons in LSE and instead of creating another "hidden" buttons in LSF we create a way to use any button you need.

#### What alternative approaches were there?
Hardcode new buttons in LSF and display them by one more interface.

#### What feature flags were used to cover this change?
None. That's addition and will be used only if specific props are passed in.

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
- [ ] e2e
- [ ] integration
- [ ] unit

hlomzik committed:
LEAP-1454: Add CustomButton store to AppStore

Simple declarative store to describe custom buttons.
Plus universal handler to handle any button depending on its type.

hlomzik committed:
Define some exerpt of types for our stores

Main stores are covered with some popular props and methods:
AppStore, AnnotationStore, Annotation

hlomzik committed:
Add custom buttons to BottomBar

They will fully substitute all other buttons for now.
Later we'll add some other buttons specified in custom buttons config.

hlomzik committed:
Reuse internal button if its name is used in custom buttons

This allows to mix custom buttons with internal ones.

hlomzik committed:
Remove unused component

hlomzik committed:
Fix types in Controls.tsx

- look for Submit is primary as a fallback value, because it has no secondary
- Button now have a `mod` property from Block, should be infered, but that's for later
- Dropdown pointing to a correct component with additional property
- the rest are just added types or direct type casts

hlomzik commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -80,7 +128,8 @@ export const Controls = controlsInjector(
     );
 
-    const RejectButton = useMemo(() => {
+    // @todo memo?
```
```suggestion
```

hlomzik committed:
Comments to explain custom buttons and types

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -80,7 +132,7 @@ export const Controls = controlsInjector(
     );
 
-    const RejectButton = useMemo(() => {
+    const RejectButton = memo(({ disabled, store }: { disabled: boolean, store: MSTStore }) => {
```
Maybe it's better to move elements outside the `Controls`, 'cause I guess in this case they'll be created on each render.

hlomzik replied:
Only functions created at each render, which is usual sacrifice nowadays. These two buttons don't have external dependencies, but others do, so for now I'd leave them as is to be more bulletproof, but I have a draft of new Controls which will improve this situation of course

Gondragos replied:
It's not about functions, it's about memo.
https://codesandbox.io/p/sandbox/react-memo-example-forked-f8yjgc?workspaceId=91ee8eb7-598d-4294-8b94-c2ca887b1c01

bmartel replied:
I agree with what @Gondragos mentioned here. The `memo()` is for component declaration, as it memoizes it based on the props, such that the usage of it would not re-render unless the props had reason to change. But because this is defined not in the module space, but within another component, it will be the wrong layer of memoization. Because it is within another component, it would have to use a hook `useMemo` to memoize the function declaration, not the return value of calling the function itself. The difference here would mean that depending on how it is used, the value of the `RejectButton` reference when rendered would cause other components to also re-render if they are memoized, because the reference is recreated each time. Or of course if there are no scope dependencies the best way is to just lift this out to a module level component, which you can decide at that level if `memo()` is still needed or not.

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -102,12 +154,10 @@ export const Controls = controlsInjector(
+    });
 
-      buttons.push(
+    const AcceptButton = memo(({ disabled, history, store }: { disabled: boolean, history: any, store: MSTStore }) => {
```
The same as for `RejectButton` https://github.com/HumanSignal/label-studio/pull/6411/files#r1768908443

hlomzik committed:
Move buttons to a separate file; fix disabled state

`buttons.tsx` to store most of the buttons, so logic will be more readable in `Controls.tsx`.
Buttons which can require a comment use handler wrapper.
`disabled` is added to custom buttons as well to block them during action handling as all other buttons.

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11013541980)
> Successfully merged:  create mode 100644 label_studio/ml_models/migrations/0012_alter_thirdpartymodelversion_provider.py

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1454/flexible-reject-buttons'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11013541980

hlomzik committed:
Fix linting

hlomzik committed:
Fix buttons to react on external changes with `observer()`
